Register independent vars on each install type

yrobla opened the pull request:
When jenkins install is called, both tasks are visited. And
it registers result of jenkins_install_package twice, causing
it to have a wrong value, and don't install jenkins users, causing
security to don't be configure and plugins to fail.

It's better to add different vars for each task, and trigger
the users install in case of some of the two changed.

geerlingguy commented:
> When jenkins install is called, both tasks are visited.

What's the use case here? I'm typically deploying to one server, or many servers with the same OS (e.g. all are Red Hat, all are Debian, etc.).

Is this a case where you have one playbook building a bunch of Jenkins servers and some are running Ubuntu, some CentOS, etc.?

Because in the former case, `jenkins_install_package` would only ever be set once, by either the RHEL or Debian included install.

yrobla commented:
I'm just running the playbook locally, on localhost. I know the task is run only once, but the var is stored in both cases. Look at some debug i added to try to detect the problem:

TASK [geerlingguy.jenkins : Ensure Jenkins is installed.] **********************
changed: [jenkins_master] => {"changed": true, "msg": "", "rc": 0, "results": ["Loaded plugins: fastestmirror\nLoading mirror speeds from cached hostfile\n * base: mirrors.evowise.com\n * epel: download-i2.fedoraproject.org\n * extras: mirrors.seas.harvard.edu\n * updates: mirror.metrocast.net\nResolving Dependencies\n--> Running transaction check\n---> Package jenkins.noarch 0:2.7.1-1.1 will be installed\n--> Finished Dependency 
Resolution\n\nDependencies Resolved\n\n================================================================================\n Package           Arch             Version             Repository         Size\n================================================================================\nInstalling:\n jenkins           noarch           2.7.1-1.1           jenkins            65 M\n\nTransaction Summary\n=============================================
 ===================================\nInstall  1 Package\n\nTotal download size: 65 M\nInstalled size: 66 M\nDownloading packages:\nRunning transaction check\nRunning transaction test\nTransaction test succeeded\nRunning transaction\n  Installing : jenkins-2.7.1-1.1.noarch                                     1/1 \n  Verifying  : jenkins-2.7.1-1.1.noarch                                     1/1 \n\nInstalled:\n  jenkins.noarch 0:2.7.1-1.1       
                                              \n\nComplete!\n"]}
 TASK [geerlingguy.jenkins : debug] *********************************************
 ok: [jenkins_master] => {
     "jenkins_install_package": {
         "changed": true, 
         "msg": "", 
         "rc": 0, 
         "results": [
             "Loaded plugins: fastestmirror\nLoading mirror speeds from cached hostfile\n * base: mirrors.evowise.com\n * epel: download-i2.fedoraproject.org\n * extras: mirrors.seas.harvard.edu\n * updates: mirror.metrocast.net\nResolving Dependencies\n--> Running transaction check\n---> Package jenkins.noarch 0:2.7.1-1.1 will be installed\n--> Finished Dependency Resolution\n\nDependencies Resolved\n\n========================================
 ========================================\n Package           Arch             Version             Repository         Size\n================================================================================\nInstalling:\n jenkins           noarch           2.7.1-1.1           jenkins            65 M\n\nTransaction Summary\n================================================================================\nInstall  1 Package\n\nTotal download 
 size: 65 M\nInstalled size: 66 M\nDownloading packages:\nRunning transaction check\nRunning transaction test\nTransaction test succeeded\nRunning transaction\n  Installing : jenkins-2.7.1-1.1.noarch                                     1/1 \n  Verifying  : jenkins-2.7.1-1.1.noarch                                     1/1 \n\nInstalled:\n  jenkins.noarch 0:2.7.1-1.1                                                    \n\nComplete!\n"
         ]
     }
 }
 TASK [geerlingguy.jenkins : Ensure dependencies are installed.] ****************
 skipping: [jenkins_master] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
 TASK [geerlingguy.jenkins : Add Jenkins apt repository key.] *******************
 skipping: [jenkins_master] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
 TASK [geerlingguy.jenkins : Add Jenkins apt repository.] ***********************
 skipping: [jenkins_master] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
 TASK [geerlingguy.jenkins : Ensure Jenkins is installed.] **********************
 skipping: [jenkins_master] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}
 TASK [geerlingguy.jenkins : debug] *********************************************
 ok: [jenkins_master] => {
     "jenkins_install_package": {
         "changed": false, 
         "skip_reason": "Conditional check failed", 
         "skipped": true
     }
 }

Look at the debug of jenkins_install_package. See that the first time, runs for RHEL, and it collects the result from the package install. But then runs for debian, and it just skips the task but stores the result anyway, with a changed false, and causing the admin user to don't be installed, security config skipped and plugins failing due to invalid parameter --username

geerlingguy commented:
Hmm... that seems like a bug with Ansible—if the included file is skipped, it shouldn't `register` any variable value.

What version of Ansible are you running currently? I'll do a little digging on my end too.

geerlingguy commented:
Or... hmm... https://github.com/ansible/ansible/issues/15710

Apparently my understanding of skipped tasks and `register` is completely incorrect :P

yrobla commented:
i install from pip
i have ansible 2.1.1.0


2016-08-01 19:45 GMT+02:00 Jeff Geerling <notifications@github.com>:

> Hmm... that seems like a bug with Ansible—if the included file is skipped,
> it shouldn't register any variable value.
>
> What version of Ansible are you running currently? I'll do a little
> digging on my end too.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/geerlingguy/ansible-role-jenkins/pull/72#issuecomment-236653338>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACaeLXP0PhPwgen9iOtSfVH01C7Z8DJ_ks5qbjCjgaJpZM4JZkCq>
> .
>



-- 
Yolanda Robla
email: info@ysoft.biz
teléfono: 605641639
